arangod/IResearch/IResearchDocument.h

```diff
@@ -220,6 +220,62 @@ class FieldIterator {
   bool _isDBServer;
 }; // FieldIterator
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief allows to iterate over the provided VPack accoring the specified
+///        IResearchInvertedIndexMeta
+////////////////////////////////////////////////////////////////////////////////
+class InvertedIndexFieldIterator {
+ public:
+  // must match interface of FieldIterator to make usable template insert implementation
+  Field const& operator*() const noexcept { return _value; }
+
+  InvertedIndexFieldIterator& operator++() {
+    next();
+    return *this;
+  }
+
+  // we don't need trx as we don't index the _id attribute.
+  // but we require it here just to match signature of "FieldIterator" in general
+  explicit InvertedIndexFieldIterator(arangodb::transaction::Methods&, irs::string_ref collection, IndexId indexId);
+
+  bool valid() const noexcept { return _fieldsMeta && _begin != _end;}
+
+  void reset(velocypack::Slice slice,
+             InvertedIndexFieldMeta const& fieldsMeta) {
+    _slice = slice;
+    _fieldsMeta = &fieldsMeta;
+    _begin = _fieldsMeta->_fields.data() - 1;
```

Comment:
Looks like we are intentionally pointing to a potentially invalid memory address.
This will work if we always do `++_begin` before dereferencing it, but that needs to be remembered throughout the code. Additionally, with that setup `_begin != _end` even if `_fieldsMeta->_fields.empty()` (not sure if the latter is possible, but it feels a bit strange).

Reply:
Yes, this is intentionally  for simplifying `next` implementation. So it could always  increment `_begin` without check if it is initial step or not. But you are right - we need an assertion that `_fieldsMeta->_fields.empty()` is false (Index factory forbids to create index with empty fields list)